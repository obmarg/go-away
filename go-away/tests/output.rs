#![allow(dead_code)]

use insta::assert_snapshot;

use go_away::{TypeMetadata, TypeRegistry};

#[derive(TypeMetadata)]
struct MyData {
    field_one: String,
    nested: Nested,
}

#[test]
fn test_struct_output() {
    let mut registry = TypeRegistry::new();
    MyData::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output(registry));
}

#[derive(TypeMetadata)]
#[serde(tag = "type", content = "data")]
enum NewTypeEnum {
    OptionOne(One),
    OptionTwo(Two),
}

#[derive(TypeMetadata)]
struct Nested {
    #[serde(rename = "some_other_name")]
    a_string: String,
    an_int: i64,
    fulfilment_type: FulfilmentType,
}

#[derive(TypeMetadata)]
enum FulfilmentType {
    Delivery,
    Collection,
}

#[derive(TypeMetadata)]
struct One {
    x: f32,
}

#[derive(TypeMetadata)]
struct Two {
    y: bool,
}

#[test]
fn test_newtype_enum() {
    let mut registry = TypeRegistry::new();
    NewTypeEnum::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output(registry));
}

#[derive(TypeMetadata)]
#[serde(tag = "type", content = "data")]
enum StructEnum {
    OptionOne { x: String, y: i32 },
    OptionTwo { foo: String, bar: Nested },
}

#[test]
fn test_struct_enum() {
    let mut registry = TypeRegistry::new();
    StructEnum::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output(registry));
}

#[derive(TypeMetadata)]
#[serde(tag = "type")]
enum InternallyTaggedTupleEnum {
    One(One),
    Two(Two),
}

#[test]
fn test_internally_tagged_tuple_enum() {
    let mut registry = TypeRegistry::new();
    InternallyTaggedTupleEnum::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output(registry));
}
