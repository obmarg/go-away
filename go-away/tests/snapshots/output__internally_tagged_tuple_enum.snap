---
source: go-away/tests/output.rs
expression: "go_away::registry_to_output(registry)"

---
type InternallyTaggedTupleEnum struct {
	*One
	*Two
}

func (self *InternallyTaggedTupleEnum) MarshalJSON() ([]byte, error) {
	if self.One != nil {
		return json.Marshal(struct{
			Tag string `json:"type"`
			One
		}{
			Tag: "One",
			One: *self.One
		})
	} else 	if self.Two != nil {
		return json.Marshal(struct{
			Tag string `json:"type"`
			Two
		}{
			Tag: "Two",
			Two: *self.Two
		})
	} else 	{
		return json.Marshal(nil)
	}
}
func (self *InternallyTaggedTupleEnum) UnmarshalJSON(data []byte) error {
	temp := struct{
		Tag string `json:"type"`
	}{}
	if err := json.Unmarshal(data, &temp); err != nil {
		return err
	}
	if temp.Tag == "One" {
		var rv One
		if err := json.Unmarshal(data, &rv); err != nil {
			return err
		}
		self.One = &rv
		self.Two = nil
	} else if temp.Tag == "Two" {
		var rv Two
		if err := json.Unmarshal(data, &rv); err != nil {
			return err
		}
		self.Two = &rv
		self.One = nil
	} else {
		return errors.New("Unknown type tag")
	}
	return nil
}
type Two bool
type One float64

